driver: ethernet: support for W6100 Ethernet controller

ParthSanepara opened the pull request:
This PR introduces the initial implementation of driver files for the WIZnet Ethernet IC W6100.

this is my first attempt at driver development, I have created this as a draft PR to seek feedback and suggestions for improvement. Reviews and guidance would be greatly appreciated to refine and enhance this work.

PR: #83029 

Thanks,

ParthSanepara committed:
driver: ethernet: support for W6100 Ethernet controller

add initial level of ethernet controller from wizznet IC W6100
fix compilation error

Signed-off-by: Parth Sanepara <parthsanepara@gmail.com>

TerryGeng commented:
Hi! I realized that most of the W6100 code you have written coincides with the existing W5500 code. I guess maybe that is because you are also using W6100 in MACRAW mode (the biggest difference between W6100 and W5500 is the support for IPv6 offload sockets). Is that the case, or maybe I have missed an important point?

ParthSanepara commented:
> Hi! I realized that most of the W6100 code you have written coincides with the existing W5500 code. I guess maybe that is because you are also using W6100 in MACRAW mode (the biggest difference between W6100 and W5500 is the support for IPv6 offload sockets). Is that the case, or maybe I have missed an important point?

Hi @TerryGeng , Thanks for pointing this out. Yes, I’m aware of the overlap with the W5500 code. This is my initial driver to enable the W6100, and I haven’t had the chance to test it with the dev kit yet. I still need to prepare the board files and plan to use the W6100-EVB-Pico for testing.

You’re correct that the W6100 supports both IPv4 and IPv6 MACRAW mode. My plan is to start with IPv4 testing, and once that’s stable, I’ll work on enhancing the driver to support all features.

This is my first time developing a driver here, so any suggestions or ideas on handling IPv4 /IPv6 and MACRAW sockets?  it would be very helpful.

Thanks again for your review!

ParthSanepara committed:
Merge branch 'zephyrproject-rtos:main' into eth_wizznet_w6100

ParthSanepara committed:
driver: ethernet: support for W6100 Ethernet controller

add initial level of ethernet controller from wizznet IC W6100
fix Compliance error in Kconfig file

Signed-off-by: Parth Sanepara <parthsanepara@gmail.com>

TerryGeng commented:
I quickly compared the documents for W5500 and W6100. If I got it right, I think for W6100, the control flow in MACRAW mode is basically the same as the W5500 chip. But there are substantial changes in the register addresses. I would guess, once you replace the register addresses of the old W5500 driver, it would work right away.

There are substantial differences regarding the chip functionalities (IPv6, socket-less commands, etc.). But if you don't really use socket offload interface, they don't really matter that much to you I assume... In that regard, W6100 is not doing anything better than W5500 seemingly to me (other than W6100 being more difficult to procure at this moment).

sramrajkar commented:
@TerryGeng One thing to point out is W5500 does not have auto MDIX which the W6100 chip has. This one of the benefits of using this chip over W5500. One good use case of have auto MDIX would be for ethernet rings.

TerryGeng commented:
I would be more interested in getting the socket offload implemented for W6100S.

I had an implementation for W5500 that is pending: https://github.com/zephyrproject-rtos/zephyr/pull/83413

I think it's a better idea that we don't duplicate the code, but try to make W5500 and W6100S share most of the code if possible.
